src/goto-programs/goto_contractor.h

```diff
@@ -0,0 +1,239 @@
+  std::string var_name[MAX_VAR];
+  symbol2tc symbols[MAX_VAR];
+
+  CspMap()
+  {
+    intervals.resize(MAX_VAR);
+  }
+  int add_var(std::string name, symbol2t symbol)
+  {
+    if(find(name) == NOT_FOUND && n < MAX_VAR)
+    {
+      symbols[n] = symbol;
+      var_name[n] = name;
+
+      ///keep for later. set initial intervals
+      //auto w = symbol.type->get_width();
+      /*if(is_signedbv_type(symbol.type))
+        add_interval(-pow(2,w-1),pow(2,w-1)-1,n);
+      else
+        add_interval(0,pow(2,w)-1,n);*/
+      n++;
+      return n - 1;
+    }
+    return NOT_FOUND;
+  }
+  void add_interval(double lb, double ub, int index)
+  {
+    intervals[index] = Interval(lb, ub);
+  }
+  void update_lb_interval(double lb, int index)
+  {
+    add_interval(lb, intervals[index].ub(), index);
+  }
+  void update_ub_interval(double ub, int index)
+  {
+    add_interval(intervals[index].lb(), ub, index);
+  }
+  int find(std::string name)
+  {
+    for(int i = 0; i < n; i++)
+      if(var_name[i] == name)
+        return i;
+
+    return NOT_FOUND;
+  }
+  int size()
+  {
+    return n;
+  }
+
+private:
+  int n = 0;
+  Variable *x;
+};
+
+class goto_contractort : public goto_functions_algorithm
+{
+  // +constructor
+
+public:
+  /**
+   * This constructor will run the goto-contractor procedure.
+   * it will go through 4 steps.
+   * First is parsing the properties.
+   * Second, parsing the intervals.
+   * Third, applying the contractor.
+   * Fourth, inserting assumes in the program to reflect the contracted intervals.
+   * @param _goto_functions
+   * @param _message_handler
+   */
+  goto_contractort(
+    goto_functionst &_goto_functions,
+    const messaget &_message_handler)
+    : goto_functions_algorithm(_goto_functions, true)
+  {
+    message_handler = _message_handler;
+    initialize_main_function_loops();
+    if(function_loops.size())
+    {
+      map = new CspMap();
+      vars = new Variable(MAX_VAR);
+      message_handler.status(
+        "1/4 - Parsing asserts to create CSP Constraints.");
+      get_constraints(_goto_functions);
+      if(constraint == nullptr)
+      {
+        message_handler.status(
+          "Constraint expression not supported. Aborting goto-contractor");
+        return;
+      }
+      message_handler.status(
+        "2/4 - Parsing assumes to set values for variables intervals.");
+      get_intervals(_goto_functions);
+      message_handler.status("3/4 - Applying contractor.");
+      auto new_intervals = contractor();
+      if(new_intervals == map->intervals)
+      {
+        message_handler.status(
+          "4/4 - Intervals remained unchanged. No assumes will be inserted");
+      }
+      else
+      {
+        message_handler.status("4/4 - Inserting assumes.");
+        insert_assume(_goto_functions, new_intervals);
+      }
+    }
+  }
+  /**
+   * This destructor that will delete all objects allocated of type Function
+   * and NumConstraint
+   */
+  ~goto_contractort()
+  {
```

Comment:
can you do something like:

for( auto c : list)
  delete(p);
list.clear();

I don't think you need to check if p is not a nullptr before calling the destructor, at least you don't need it for free.